contracts/NonfungiblePositionManager.sol

```diff
@@ -146,37 +145,55 @@ contract NonfungiblePositionManager is INonfungiblePositionManager, BaseLiquidit
+
+        delete tokenRange[tokenId];
         // Burn the token.
         _burn(tokenId);
     }
 
-    // TODO: Bug - Positions are overrideable unless we can allow two of the same users to have distinct positions.
-    function _beforeTokenTransfer(address from, address to, uint256 tokenId) internal override {
-        TokenPosition storage tokenPosition = tokenPositions[tokenId];
-        LiquidityRangeId rangeId = tokenPosition.range.toId();
-        Position storage position = positions[from][rangeId];
-
-        // transfer position data to destination
-        positions[to][rangeId] = position;
-        delete positions[from][rangeId];
-
-        // update token position
-        tokenPositions[tokenId] = TokenPosition({owner: to, range: tokenPosition.range, operator: address(0x0)});
+    function _modifyLiquidity(LiquidityRange memory range, int256 liquidityChange, bytes32 salt, bytes memory hookData)
+        internal
+        returns (BalanceDelta liquidityDelta, BalanceDelta totalFeesAccrued)
+    {
+        (liquidityDelta, totalFeesAccrued) = manager.modifyLiquidity(
+            range.poolKey,
+            IPoolManager.ModifyLiquidityParams({
+                tickLower: range.tickLower,
+                tickUpper: range.tickUpper,
+                liquidityDelta: liquidityChange,
+                salt: salt
+            }),
+            hookData
+        );
     }
 
-    // override ERC721 approval by setting operator
-    function _approve(address spender, uint256 tokenId) internal override {
-        tokenPositions[tokenId].operator = spender;
+    function _validateBurn(uint256 tokenId) internal {
+        bytes32 positionId = getPositionIdFromTokenId(tokenId);
+        uint128 liquidity = manager.getPositionLiquidity(tokenRange[tokenId].poolKey.toId(), positionId);
+        if (liquidity > 0) revert PositionMustBeEmpty();
     }
 
-    function getApproved(uint256 tokenId) public view override returns (address) {
-        require(_exists(tokenId), "ERC721: approved query for nonexistent token");
-
-        return tokenPositions[tokenId].operator;
+    function getPositionIdFromTokenId(uint256 tokenId) public view returns (bytes32 positionId) {
+        LiquidityRange memory range = tokenRange[tokenId];
+        bytes32 salt = bytes32(tokenId);
+        int24 tickLower = range.tickLower;
+        int24 tickUpper = range.tickUpper;
+        address owner = address(this);
+
+        // TODO: reorganize this into library on core?
+        // positionId = keccak256(abi.encodePacked(owner, tickLower, tickUpper, salt))
+        /// @solidity memory-safe-assembly
+        assembly {
+            mstore(0x26, salt) // [0x26, 0x46)
+            mstore(0x06, tickUpper) // [0x23, 0x26)
+            mstore(0x03, tickLower) // [0x20, 0x23)
+            mstore(0, owner) // [0x0c, 0x20)
+            positionId := keccak256(0x0c, 0x3a) // len is 58 bytes
+            mstore(0x26, 0) // rewrite 0x26 to 0
+        }
```

Comment:
This can be a view function outside the main contract IMO.. ie we can have a LiquidityRouterState.sol that people import that just uses StateLibrary.sol - we can leave for now but I prefer all view functions to live outside the main contract